server/app/database.py

```diff
@@ -51,16 +51,26 @@ def compile(query: sa.sql.elements.ClauseElement) -> str:
     "measurements",
     metadata,
     sa.Column(
-        "node_identifier",
-        sa.String(length=validation.Limit.MEDIUM),
+        "sensor_identifier",
+        sa.String(length=constants.Limit.MEDIUM),
         sa.ForeignKey(
-            CONFIGURATIONS.columns.node_identifier,
+            CONFIGURATIONS.columns.sensor_identifier,
             onupdate="CASCADE",
-            ondelete="CASCADE",
+            ondelete="CASCADE",  # cascade is so fucking sexy
         ),
         nullable=False,
     ),
     sa.Column("measurement_timestamp", sa.Integer, nullable=False),
     sa.Column("receipt_timestamp", sa.Integer, nullable=False),
+    # TODO implement as JSON for maximum flexibility?
     sa.Column("value", sa.Integer, nullable=False),
```

Comment:
Do you mean, using a JSON in the value column? Could be really cool because then the sensors can send whatever data and just have to fulfil our meta-data requirements (`measurement_timestamp`, etc.).

This makes it impossible to do an aggregation in native SQL, but I don't think that this is a big drawback.

We have to do our own validation in the JSON content in the `data` field submitted by the node.

Reply:
That's it. I'm already using that as the type for the sensor configuration. There it's easier because we don't have to work on the data, other than reading values.

For the measurements, I still have to see what the better option is: having a table for every type of measurement or using JSON. I'm trying to keep both the structure of the configuration and the measurement data as flexible as possible.

(You can in fact run some operations on JSON in PostgreSQL, I'm not sure yet, to what extent though, but it seems pretty flexible.)

> We have to do our own validation in the JSON content in the data field submitted by the node.

I don't really understand what you mean by that. In any case, we can usually validate very basic things, like limits on the size of the whole thing, or min/max values for integers, but it's pretty difficult to say, yes, this (e.g.) `CO2` value is valid in a way that makes sense.

Reply:
> [...] having a table for every type of measurement or using JSON.

JSON definitely has less redundancy! I think it is better with JSON fields than with separate tables.

> I don't really understand what you mean by that. In any case, we can usually validate very basic things, like limits on the size of the whole thing, or min/max values for integers, but it's pretty difficult to say, yes, this (e.g.) CO2 value is valid in a way that makes sense.

Only structure, not if the values make sense. Just not "any text". You would do that anyway :D

Reply:
> I think it is better with JSON fields than with separate tables.

I think so as well, but I have to see how it works out. I believe that using "arbitrary" JSON values as:

1. measurements
2. sensor configuration
3. sensor health updates

could bring lots of nice flexibility. The downside is that the validation is a bit more tricky and that we need to be more careful about what we do with the values. For example, the sensor would need to be able to handle invalid configurations, because the server can't really validate them except for structure.

I think we try it out and see where we end up 😄 I'll definitely validate:

- min/max values for int and float 
- max length for strings
- JSON can't be nested
- max number of JSON fields